src/test/java/echoserver/EchoServerTest.java

```diff
@@ -1,38 +1,90 @@
 package echoserver;
 
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.concurrent.Executor;
 
 import static org.junit.jupiter.api.Assertions.*;
 
 class EchoServerTest {
+    Echoable echoer;
+    TestConnection firstConnection;
+    TestConnection secondConnection;
+    TestConnection thirdConnection;
+    ArrayList<TestConnection> clients;
+    FakeListener listener;
+    SerialExecutor executor;
+    errorLogger logger;
+    EchoServer echoServer;
+
+    @BeforeEach
+    void setUp() {
+        firstConnection = new TestConnection();
+        secondConnection = new TestConnection();
+        thirdConnection = new TestConnection();
+        clients = new ArrayList<TestConnection>(Arrays.asList(firstConnection, secondConnection, thirdConnection));
+        listener = new FakeListener(clients);
+        executor = new SerialExecutor();
+        logger = new errorLogger();
+    }
+
+    @Test
+    void itHandlesMultipleConnections() throws Exception {
+        echoer = new EchoOnlyOnce();
+        echoServer = new EchoServer(listener, executor, echoer, logger);
+
+        echoServer.start();
+
+        assertEquals(Arrays.asList(firstConnection, secondConnection, thirdConnection),listener.connectedClients);
+    }
+
+    @Test
+    void itExecutesTheRunnableForEachConnection() throws Exception {
+        echoer = new EchoOnlyOnce();
+        echoServer = new EchoServer(listener, executor, echoer, logger);
+
+        echoServer.start();
+
+        assertEquals(Arrays.asList("executed", "executed", "executed"), executor.executed);
+    }
+
     @Test
-    void itConnectsToClientsInOrderReceived() throws Exception {
-        FakeEchoClient echoClient = new FakeEchoClient();
-        TestConnection firstClient = new TestConnection();
-        TestConnection secondClient = new TestConnection();
-        TestConnection thirdClient = new TestConnection();
-        ArrayList<TestConnection> clients = new ArrayList<TestConnection>(Arrays.asList(firstClient, secondClient, thirdClient));
-        FakeListener listener = new FakeListener(clients);
-        EchoServer echoServer = new EchoServer(listener, echoClient);
+    void itEchoesEachConnection() throws Exception {
+        echoer = new EchoOnlyOnce();
+        echoServer = new EchoServer(listener, executor, echoer, logger);
 
         echoServer.start();
 
-        assertEquals(Arrays.asList(firstClient, secondClient, thirdClient),listener.connectedClients);
+        assertEquals("echoechoecho", firstConnection.writes + secondConnection.writes + thirdConnection.writes);
```

Comment:
👍 

It seems like you are always working with these connections together. I wonder if it would be beneficial to work with them as a collection instead of 3 separate objects. Might be something to look into?

Reply:
Oh looks like you already did this.